adapters/ucfunnel/ucfunnel.go

```diff
@@ -0,0 +1,137 @@
+package ucfunnel
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"net/http"
+)
+
+type UcfunnelAdapter struct {
+	URI string
+}
+
+func NewUcfunnelBidder(client *http.Client, endpoint string) *UcfunnelAdapter {
+	return &UcfunnelAdapter{
+		URI: endpoint}
+}
+
+func (a *UcfunnelAdapter) Name() string {
+	return "ucfunnel"
+}
+
+func (a *UcfunnelAdapter) SkipNoCookies() bool {
+	return false
+}
+
+func (a *UcfunnelAdapter) MakeBids(internalRequest *openrtb.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	var errs []error
+	var bidResp openrtb.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	var bidReq openrtb.BidRequest
+	if err := json.Unmarshal(externalRequest.Body, &bidReq); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(5)
+	for _, sb := range bidResp.SeatBid {
+		for i := range sb.Bid {
+			bidType := getBidType(bidReq, sb.Bid[i].ImpID)
+			if (bidType == openrtb_ext.BidTypeBanner) || (bidType == openrtb_ext.BidTypeVideo) {
+				b := &adapters.TypedBid{
+					Bid:     &sb.Bid[i],
+					BidType: bidType,
+				}
+				bidResponse.Bids = append(bidResponse.Bids, b)
+			}
+		}
+	}
+	return bidResponse, errs
+}
+
+func (a *UcfunnelAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	errs := make([]error, 0, len(request.Imp))
+
+	// If all the requests were malformed, don't bother making a server call with no impressions.
+	if len(request.Imp) == 0 {
+		return nil, errs
+	}
+
+	partnerId := getPartnerId(request)
+	if len(partnerId) == 0 {
+		return nil, []error{}
```

Comment:
Instead of throwing an empty error list, we could probably use whatever unmarshall error `getPartnerId(request *openrtb.BidRequest)` delivers. The code as is, is already checking for the lenght of `PartnerID` inside `getPartnerId` anyways (line 111)
```
 74   func (a *UcfunnelAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
 75       errs := make([]error, 0, len(request.Imp))
 76  
 77   *--  5 lines: If all the requests were malformed, don't bother making a server call with no impressions.-----------------------------
 82 -     partnerId := getPartnerId(request)
 83 -     if len(partnerId) == 0 {
 84 -         return nil, []error{}
    +     if partnerId, err := getPartnerId(request); err != nil {
    +         return nil, []error{err}
 85       }
 86  
 87       reqJSON, err := json.Marshal(request)
 88       if err != nil {
 89           errs = append(errs, err)
 90           return nil, errs
 91       }
 92  
 93   *-- 10 lines: headers := http.Header{}------------------------------------------------------------------------------------------------
103   }
104  
105 - func getPartnerId(request *openrtb.BidRequest) string {
    + func getPartnerId(request *openrtb.BidRequest) (string, error) {
106       var ext ExtBidderUcfunnel
107       err := json.Unmarshal(request.Imp[0].Ext, &ext)
108       if err != nil {
109 -         return ""
    +         return "", err
110       }
111       if len(ext.Bidder.PartnerId) == 0 || len(ext.Bidder.AdUnitId) == 0 {
112 -         return ""
    +         return "", fmt.Errorf("Incorrect lenght of bidder fields: len(ext.Bidder.PartnerId) = %d; len(ext.Bidder.AdUnitId) = %d; \n", ext.Bidder.PartnerId, ext.Bidder.AdUnitId) 
113       }
114  
115 -     return ext.Bidder.PartnerId
    +     return ext.Bidder.PartnerId, nil
116   }
adapters/ucfunnel/ucfunnel.go
```